### PR TITLE
Add various mechanics tests

### PR DESCRIPTION
**Pull request by DaWoblefet**
Hooray for more mechanics tests! This PR introduces about 30 new skipped mechanics tests as well as about 15 new passing tests. All bug reports in the thread up to post 1084 are accounted for. In addition, the [mechanics bugs Kanban board](https://github.com/smogon/pokemon-showdown/projects/3) is now synchronized with the failing tests in the repository, so you can more easily view these as cards if you find that easier to read.

### CODE REVIEW

File: test/sim/items/whiteherb.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			{species: 'wynaut', item: 'whiteherb', moves: ['sleeptalk']},
+		]]);
+		battle.makeChoices();
+		const wynaut = battle.p2.active[0];
+		assert.false.holdsItem(wynaut);
+		assert.statStage(wynaut, 'atk', 0);
+		assert.statStage(wynaut, 'spa', 0);
+	});
+
+	it.skip('should activate after Abilities that boost stats on KOs', function () {
```
Comment (MacChaeger) [inline test/sim/items/whiteherb.js:27]: Didn’t the bug report say that this is what already happens on sim and that in-game, White Herb should go first? Also, this test is duplicated below.

Comment (DaWoblefet) [reply test/sim/items/whiteherb.js:27]: No, it should go like this (Soul-Heart as an example):
Magearna uses Fleur Cannon into target
Special Attack drops to -2
Target is KOed
Soul-Heart boosts to -1
White Herb restores to 0

which is what this test does. I'll fix the duplicate though!

Comment (MacChaeger) [reply test/sim/items/whiteherb.js:27]: In that case, try un-skipping this test. I’m pretty sure this is what the sim does already.

Comment (DaWoblefet) [reply test/sim/items/whiteherb.js:27]: Nope, it fails, because the sim is currently restoring the -2 Sp. Atk drop, then applying the Ability KO boost instead of waiting.

``1) White Herb
       should activate after Abilities that boost stats on KOs:
     AssertionError [ERR_ASSERTION]: Expected p2a: Wynaut's spa at stage 0, not at 1.``

**Comment by DaWoblefet** [comment]
phew, that should be all of the Kanban board bugs that didn't have corresponding mechanics tests. Time for the other way around (skipped tests in the repo that don't have corresponding Kanban board entries)

**Comment by DaWoblefet** [comment]
Ok, this is ready for review. There's a couple more Kanban board things I'd like to clean up, but the tests in the repo should all be done.

File: test/sim/items/shellbell.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+const common = require('./../../common');
+
+let battle;
+
+describe('Shell Bell', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it.skip(`should heal from the damage against all targets of the move`, function () {
```
Comment (urkerab) [inline test/sim/items/shellbell.js:13]: But just the last hit of a multi-hit move? What about Parental Bond?

File: test/sim/misc/arceus.js [hunk 3]
```diff
@@ -3,38 +3,69 @@
...
+			{species: 'arceus', ability: 'multitype', item: 'flameplate', moves: ['burnup']},
+		], [
+			{species: 'reuniclus', moves: ['soak']},
+		]]);
+		battle.makeChoices();
+		const arceus = battle.p1.active[0];
+		assert(arceus.hasType('Fire'), 'Arceus should not change type.');
+	});
+
+	it.skip(`should use Arceus's real type for Revelation Dance`, function () {
```
Comment (urkerab) [inline test/sim/misc/arceus.js:61]: What happens if Dancer Noivern is Roosting and then copies Revelation Dance?

File: test/sim/moves/electricterrain.js [hunk 4]
```diff
@@ -95,4 +95,14 @@ describe('Electric Terrain', function () {
 		const resultMove = toID(battle.log[battle.lastMoveLine].split('|')[3]);
 		assert.equal(resultMove, 'thunderbolt');
 	});
+
+	it.skip(`should block Sleep before the move would have missed`, function () {
```
Comment (urkerab) [inline test/sim/moves/electricterrain.js:99]: This is just for status moves that cause Sleep, right? Relic Song's Sleep just gets silently blocked as normal.

File: test/sim/abilities/cursedbody.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const common = require('./../../common');
+
+let battle;
+
+describe(`Cursed Body`, function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it.skip(`should be able to trigger on Z-moves`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/cursedbody.js]: I actually wrote this test wrong, and your comment prompted me (thanks!) - Cursed Body can Disable the Z-move itself, **not** the base of the Z-move. Disable can't Disable the Z-move, but Cursed Body can, because yeah.

I am not familiar offhand with every mechanic listed but I believe all are accurate (save the Cursed Body one that I have corrected now); I will follow up after further testing on a couple of them.

File: test/sim/abilities/magician.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+const common = require('./../../common');
+
+let battle;
+
+describe('Magician', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it.skip(`should not steal Weakness Policy on super-effective hits`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/magician.js:13]: It has to do with the order of checks done _after_ move success, which isn't well-documented yet. It's one of the things on my to-do list with the new USUM research. Suction Cups should be a check done _before_ move success though (if we're talking Suction Cups vs. Roar). 

Sort of like how accuracy works with prioritizing abilities > items, Weakness Policy should trigger using the same event, but after things like Flame Body / Rough Skin / Stamina. That event is sandwiched between things like Core Enforcer (before) and Jaboca Berry (after). Magician's timing is is alongside things like Beast Boost / Eject Button / Berserk, which is much later. Without the checks done _after_ move success properly sorted out yet though, this probably can be delayed.

File: test/sim/abilities/cursedbody.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/cursedbody.js]: So, if Cursed Body can disable the Z-move itself, what about a move called by Dancer?

File: test/sim/abilities/pressure.js [hunk 7]
```diff
@@ -85,6 +85,50 @@ describe('Pressure', function () {
...
+			{species: 'wynaut', moves: ['assist']},
+			{species: 'yveltal', moves: ['darkpulse']},
+		], [
+			{species: 'absol', ability: 'pressure', moves: ['sleeptalk']},
+		]]);
+		battle.makeChoices('move assist', 'auto');
+		assert.equal(battle.p1.active[0].getMoveData(Dex.getMove('assist')).pp, 30);
+	});
+
+	it.skip(`Sticky Web should not deduct PP from Pressure (only entry hazard to do so)`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/pressure.js:120]: Right. Imprison, Snatch, Spikes, Toxic Spikes, and Stealth Rock are all field-target moves explicitly assigned a "Pressure also works on these!", but Sticky Web didn't make it in the list :(

File: test/sim/moves/electricterrain.js (hunk 4 above)
Comment (DaWoblefet) [reply test/sim/moves/electricterrain.js:99]: Right, only moves that Electric Terrain would "loudly" block.

File: test/sim/abilities/magician.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/magician.js:13]: Actually I was probably thinking about Suction Cups vs. Dragon Tail on a Dragon-type with Weakness Policy, rather than against Roar, which gets processed differently.

File: test/sim/abilities/pressure.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/pressure.js:120]: (The sim actually considers Spikes, Toxic Spikes and Stealth Rock to be foe side-target moves rather than field-target moves, which is why the sim automatically calculates Pressure on those, but yes, it too has to special-case Snatch as well as Imprison.)

File: test/sim/abilities/sturdy.js [hunk 8]
```diff
@@ -56,4 +56,39 @@ describe('Sturdy', function () {
 		battle.makeChoices('move sleeptalk', 'move fusionflare');
 		assert.fainted(battle.p1.active[0]);
 	});
+
+	it.skip(`should trigger before Focus Sash`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/sturdy.js:60]: The order on cart for that event:
1. Endure
2. False Swipe / Hold Back
3. Sturdy
4. Focus Sash / Focus Band

False Swipe should be in here; I have no idea what Bide is doing here, it doesn't belong. Disguise also doesn't belong (should be earlier), but it does at least avoid the problem of Focus Sash triggering before it or something.

[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/sturdy.js:60]: Wait. so if you use False Swipe against a foe that's using Endure then the cart announces that the foe endured the hit? That's weird...

Disguise is in my list just because breaking Disguise successfully causes zero damage before Focus Sash can intervene, and Bide is in my list just because the amount of damage recorded by Bide is after the reduction from the above effects.

File: test/sim/abilities/pressure.js [hunk 9]
```diff
@@ -85,6 +85,50 @@ describe('Pressure', function () {
...
+		battle = common.gen(7).createBattle([[
+			{species: 'wynaut', item: 'darkiniumz', moves: ['darkpulse']},
+		], [
+			{species: 'absol', ability: 'pressure', moves: ['sleeptalk']},
+		]]);
+		battle.makeChoices('move darkpulse zmove', 'auto');
+		assert.equal(battle.p1.active[0].getMoveData(Dex.getMove('darkpulse')).pp, 22);
+	});
+
+	it.skip(`should deduct additional PP from submoves that target Pressure`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/pressure.js:109]: > So, you're saying that if you use a self-targeting move that calls moves, and that move calls a move that then targets an opposing Pressure user, then additional PP is deducted from the base move?

Yes. The move chosen by Sleep Talk, Metronome, etc. are resolved prior to target being determined. In fact, Pressure is how we know that's the case.

> What about Nature Power? Currently I think the sim might already deduct additional PP.

Yeah, Pressure should deduct an additional PP if targeted by a move called from Nature Power.

> Also, what happens to Pledge moves?

Both the interrupted Pledge move and Pledge move proper consume 2 PP when used into Pressure.

File: test/sim/abilities/receiver.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+const common = require('./../../common');
+
+let battle;
+
+describe(`Receiver`, function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it.skip(`should gain a boost immediately if taking over a KO boost Ability`, function () {
```
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/receiver.js:13]: In the scenario described, it goes like this:

1. Ally Magearna faints
2. Receiver copies ally Soul-Heart
3. Soul-Heart activates on Passimian, then opposing Magearna (despite the fact that it's going to faint). Should be Speed order.
4. Opposing Magearna faints.
5. Opposing Receiver copies opposing Soul-Heart.
6. Soul-Heart activates on both new Passimian in order of Speed.

I have a video of this that I'll upload overnight. In addition, I'll write a couple more tests for this, as this isn't very obvious behavior at all. Soul-Heart can activate when you're set to faint, but only if an opponent was KOed before you were queued to be KOed - it doesn't work if two Soul-Heart are on the same side.

File: test/sim/abilities/cursedbody.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/cursedbody.js]: Great follow-up! There was one thing mistaken in this list - Spite, Eerie Spell, and G-Max Depletion all can work on Max Moves, and they reveal the base move when doing so, like Grudge.

File: test/sim/misc/arceus.js (hunk 3 above)
Comment (DaWoblefet) [reply test/sim/misc/arceus.js:61]: That Noivern uses a Dragon-type Revelation Dance (verified into Color Change). Should a Roosting Tornadus attempt a similar task, its Revelation Dance will be Normal-type.

File: test/sim/items/shellbell.js (hunk 2 above)
Comment (DaWoblefet) [reply test/sim/items/shellbell.js:13]: Actually, Parental Bond and Shell Bell should also consider all of their hits. I'll add tests for those too! I imagine that means Shell Bell behavior changed at some point after Gen 7 (I tested in 7 just to make sure I had an accurate Parental Bond).

**Comment by DaWoblefet** [comment]
Thanks to additional insight from @urkerab, I have added a few additional tests to clarify behavior and document new tested behavior.

File: test/sim/abilities/pressure.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/pressure.js:120]: Bah, I used the wrong language with "field-target"; they of course only target one side. But yeah, it's a hardcode either way probably.

File: test/sim/abilities/sturdy.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/sturdy.js:60]: > Wait. so if you use False Swipe against a foe that's using Endure then the cart announces that the foe endured the hit? That's weird...

Yep, that's right.

**Review by Zarel — APPROVED** [review]
Looks good! I'll give a few days in case urkerab has more comments?

File: test/sim/abilities/pressure.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/pressure.js:109]: The Nature Power one is confusing as in Gen 6 it targets normally despite being a move that calls a submove (those normally target yourself). I don't know whether it's possible for it to call a spread move (I think it would require a double or multi battle in a volcano) as that would get very confusing for PP deduction purposes.

**Comment by urkerab** [comment]
If I have any more comments I'll just ask in Discord.

File: test/sim/abilities/pressure.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/pressure.js:109]: I don't think it'd be any more unusual than it is now. Just like Sleep Talk -> Thunderbolt can update the target from self to opponent, so Nature Power -> Lava Plame would update the targets to each Pokemon on the field. Hopefully we don't have to worry about Volcano battles anytime soon! :P

File: test/sim/abilities/magician.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/magician.js:13]: Aha, that makes more sense. Dragon Tail's switching effect is run after Weakness Policy, in the same event as things like Kee Berry or Berserk. So you would Dragon Tail, deal damage, trigger the Weakness Policy, and then be forced to switch out. But we're getting into the weeds with testing edge cases for checks done after move success; I'll add Dragon Tail vs. Weakness Policy but leave the general mechanic alone for this batch of tests.

File: test/sim/abilities/receiver.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/receiver.js:13]: Here's that video! https://www.youtube.com/watch?v=wDuCh1h6lzI

File: test/sim/items/shellbell.js (hunk 2 above)
Comment (DaWoblefet) [reply test/sim/items/shellbell.js:13]: Here's a video of Shell Bell + Rock Blast to clarify a bit: https://www.youtube.com/watch?v=_VT6QtL1KJo

File: test/sim/abilities/cursedbody.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaWoblefet) [reply test/sim/abilities/cursedbody.js]: > So, if Cursed Body can disable the Z-move itself, what about a move called by Dancer?

Yep!

**Comment by DaWoblefet** [comment]
Yep, I'm ready to go now @Zarel. Merge away!

**Comment by Zarel** [comment]
Thanks!

File: test/sim/abilities/magician.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (urkerab) [reply test/sim/abilities/magician.js:13]: (The question I was actually asking was whether Suction Cups activates to prevent the forced switch before or after Weakness Policy.)
